connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java

```diff
@@ -1170,78 +983,128 @@ public void testDuplicatedAssignmentHandleWhenTheDuplicatedAssignmentsDeleted()
-        applyAssignments(returnedAssignments);
-        memberConfigs = memberConfigs(leader, offset, assignments);
-        assignor.performTaskAssignment(leader, offset, memberConfigs, coordinator, protocolVersion);
+        performStandardRebalance();
+        assertDelay(0);
+        assertEmptyAssignment();
+
+        verifyCoordinatorInteractions();
+    }
+
+    private void performStandardRebalance() {
+        performRebalance(false, false);
+    }
+
+    private void performFailedRebalance() {
+        performRebalance(true, false);
+    }
+
+    private void performRebalanceWithMismatchedGeneration() {
+        performRebalance(false, true);
+    }
+
+    private void performRebalance(boolean assignmentFailure, boolean expectGenerationMismatch) {
+        expectGeneration(expectGenerationMismatch);
+        // Member configs are tracked by the assignor; create a deep copy here so that modifications to our own memberConfigs field
+        // are not accidentally propagated to the one used by the assignor
+        Map<String, ExtendedWorkerState> memberConfigsCopy = memberConfigs.entrySet().stream().collect(Collectors.toMap(
+                Map.Entry::getKey,
+                e -> {
+                    ExtendedWorkerState originalWorkerState = e.getValue();
+                    return new ExtendedWorkerState(
+                            originalWorkerState.url(),
+                            originalWorkerState.offset(),
+                            duplicate(originalWorkerState.assignment())
+                    );
+                }
+        ));
+        try {
+            assignor.performTaskAssignment(leader, offset, memberConfigsCopy, coordinator, protocolVersion);
+        } catch (RuntimeException e) {
+            if (assignmentFailure) {
+                RequestFuture.failure(e);
+            } else {
+                throw e;
+            }
+        }
         ++rebalanceNum;
         returnedAssignments = assignmentsCapture.getValue();
-        assertDelay(0, returnedAssignments);
-        expectedMemberConfigs = memberConfigs(leader, offset, returnedAssignments);
-        assertNoReassignments(memberConfigs, expectedMemberConfigs);
-        assertAssignment(0, 0, 0, 0, "worker1", "worker2");
+        assertNoRedundantAssignments();
+        if (!assignmentFailure) {
+            applyAssignments(leader, offset, returnedAssignments);
+        }
+    }
 
-        verify(coordinator, times(rebalanceNum)).configSnapshot();
-        verify(coordinator, times(rebalanceNum)).leaderState(any());
-        verify(coordinator, times(2 * rebalanceNum)).generationId();
-        verify(coordinator, times(rebalanceNum)).memberId();
-        verify(coordinator, times(rebalanceNum)).lastCompletedGenerationId();
+    private void expectGeneration(boolean expectMismatch) {
+        when(coordinator.generationId())
+                .thenReturn(assignor.previousGenerationId + 1)
+                .thenReturn(assignor.previousGenerationId + 1);
```

Comment:
Since we return the same value, we can remove one of these `thenReturn()`

Reply:
Ah, good call. I figured the [existing style](https://github.com/apache/kafka/blob/481cc13a132d33f23e737f88ae28a1aac135afed/connect/runtime/src/test/java/org/apache/kafka/connect/runtime/distributed/IncrementalCooperativeAssignorTest.java#L1464-L1466) was necessary since `WorkerCoordinator::generationId` would be invoked twice for every call to `expectGeneration` but it looks like Mockito handles this just fine with a single `thenReturn`.